### PR TITLE
storage: make k8s emptyDir volume creation location configurable

### LINKED ISSUE #2053: Add support for host-created EmptyDir
This is a recreation of [this issue](https://github.com/kata-containers/runtime/issues/2518) raised against Kata 1.X. A PR will be inbound shortly.

**Is your feature request related to a problem? Please describe.**

Kubernetes EmptyDirs are sometimes used as a way to share data from the host to a guest. The work done in https://github.com/kata-containers/runtime/pull/1485 drastically improved performance for most use-cases, but broke host->guest sharing of files.

**Describe the solution you'd like**

A flag in the configuration that disables guest-created EmptyDirs, allowing users to fall back to slower, host-created ones.

**Describe alternatives you've considered**

Unfortunately, there aren't really any other good solutions to the problem my org is using EmptyDir to solve. There are multi-tenancy issues with host volumes, and our use-case relies on linking container logs on the host to a logging container in a pod.

**Additional context**

I've already implemented this and my org is running a fork with a disable_guest_empty_dir flag, but I'd like to gauge upstream's interest in this. It's a very small changeset (+42 -11) and would allow folks to use EmptyDir with Kata in this way again.

**Before raising this feature request**

Have you looked at the [limitations document](https://github.com/kata-containers/documentation/blob/master/Limitations.md)?

Yes.

### PR DESCRIPTION
**Pull request by evanfoster**
This is a forward-port of the change that fixed this Kata 1.X issue: https://github.com/kata-containers/runtime/issues/2518

In that issue, folks had suggested I make this configurable via annotations. I tried to do that but I wasn't able to figure it out, so I'm opening this PR to get some suggestions about how I might implement that (or if it's still necessary).

This change introduces the `disable_guest_empty_dir` config option, which allows the user to change whether a Kubernetes emptyDir volume is created on the guest (the default, for performance reasons), or the host (necessary if you want to pass data from the host to a guest via an emptyDir).

Fixes #2053

Signed-off-by: Evan Foster <efoster@adobe.com>

### CODE REVIEW

**Comment by egernst** [comment]
Interesting -- usually an emptyDir should be... empty.  Do you have a downstream use case where you are passing certs or other special files via emptyDir?

I know a guy who has similar scenario.

**Comment by evanfoster** [comment]
Hey @egernst,

We're currently using emptyDirs as a way to pass container runtime logs for a pod on the host into a logging sidecar container running in that pod. We can't use things like HostPath, since it's unfortunately not safe for multitenant usage.

@c3d had mentioned wanting annotation support in the associated issue (#2053), and it seems like it may be possible now (unlike in Kata 1.x). I'm going to check what annotations are available when volume setup is going on. It can't hurt to have them if it's possible without adding a bunch of bloat, although I agree with your assessment. I haven't seen much of a performance difference when comparing host vs. guest-created emptyDirs using virtio-fs.

**Comment by amshinde** [comment]
@egernst The idea was to eventually just use the bind mounts. We started creating the empty volumes within the guest to have better performance in case of devicemapper specifically to support docker in docker builds. I am not sure if there was any progress comparing devicemapper performance to virtiofs. Depending on those results we can consider dropping creating empty-dir volumes within the guest altogether else go with the approach in this PR.

**Comment by jodh-intel** [comment]
Thanks for rebasing @evanfoster.

@amshinde - could you tal at this as I think you added the 1.x support?

**Comment by evanfoster** [comment]
Hey @amshinde, have you had a chance to take another look at this?

**Comment by Jakob-Naucke** [comment]
@amshinde ^

**Comment by jodh-intel** [comment]
Ping @amshinde.

**Comment by jodh-intel** [comment]
Re-ping @amshinde.

**Review by amshinde — APPROVED** [review]

**Comment by fidencio** [comment]
A head's up that may affect your PR.

In the Architecture Committee meeting from January 25th, 2022, the Architecture Committee has agreed on using the "Dismiss stale pull request approvals when new commits are pushed" configuration from GitHub.  It basically means that if your PR has been rebased or updated, the approvals given will be erased.

In order to minimize traumas due to the new approach, please, consider adding a note on the changes done before the rebase / force-push, and also pinging the reviewers for a subsequent round of reviews.

Thanks for your understanding!

Related issue: https://github.com/kata-containers/community/issues/249

**Comment by evanfoster** [comment]
I'll get these conflicts resolved on Monday.

**Review by jodh-intel — APPROVED** [review]
Thanks @evanfoster !

lgtm

ping @amshinde.

**Comment by egernst** [comment]
(rebased for Evan)

File: src/runtime/pkg/katautils/create.go
Comment (egernst) [inline src/runtime/pkg/katautils/create.go:222]: The parameter list here keeps getting worse and worse :|

Comment (fidencio) [reply src/runtime/pkg/katautils/create.go:222]: @egernst, although I share the same feeling as you do, I don't think this PR is making it that much worse.

Comment (egernst) [reply src/runtime/pkg/katautils/create.go:222]: its definitely making it about 11% worse :D

**Review by egernst — APPROVED** [review]

**Pull request merged**
